crates/core/src/db/queries/collections.rs

```diff
@@ -0,0 +1,149 @@
+                    AND auction_houses.treasury_mint = 'So11111111111111111111111111111111111111112'
+                    AND purchases.created_at >= $2
+                    AND purchases.created_at <= $3
+                GROUP BY metadata_collection_keys.collection_address
+                ORDER BY volume {order_direction}
+                LIMIT $4
+                OFFSET $5
+        ) SELECT
+            metadatas.address,
+            metadatas.name,
+            metadatas.seller_fee_basis_points,
+            metadatas.update_authority_address,
+            metadatas.mint_address,
+            metadatas.primary_sale_happened,
+            metadatas.uri,
+            metadatas.slot,
+            metadata_jsons.description,
+            metadata_jsons.image,
+            metadata_jsons.category,
+            metadata_jsons.model
+        FROM metadata_jsons, volume_table, metadatas
+        WHERE
+            volume_table.collection = metadatas.mint_address
+            AND metadatas.address = metadata_jsons.metadata_address
+    -- $1: addresses::text[]
+    -- $2: start date::timestamp
+    -- $3: end date::timestamp
+    -- $4: limit::integer
+    -- $5: offset::integer",
+        order_direction = order_direction
+    )
+}
+
+/// Query collections ordered by market cap
+///
+/// # Errors
+/// returns an error when the underlying queries throw an error
+pub fn by_market_cap(
+    conn: &Connection,
+    addresses: impl ToSql<Nullable<Array<Text>>, Pg>,
+    order_direction: OrderDirection,
+    start_date: DateTime<Utc>,
+    end_date: DateTime<Utc>,
+    limit: impl ToSql<Integer, Pg>,
+    offset: impl ToSql<Integer, Pg>,
+) -> Result<Vec<Nft>> {
+    diesel::sql_query(make_by_market_cap_query_string(order_direction))
+        .bind(addresses)
+        .bind::<Timestamp, _>(start_date.naive_utc())
+        .bind::<Timestamp, _>(end_date.naive_utc())
+        .bind(limit)
+        .bind(offset)
+        .load(conn)
+        .context("Failed to load collections by market cap")
+}
+
+fn make_by_market_cap_query_string(order_direction: OrderDirection) -> String {
+    format!(
+        r"
+        WITH market_cap_table (collection, market_cap) as (
+            SELECT metadata_collection_keys.collection_address AS collection, MIN(listings.price) * COUNT( DISTINCT metadata_collection_keys.metadata_address) AS market_cap
+            FROM listings
+            INNER JOIN metadatas on (listings.metadata = metadatas.address)
+            INNER JOIN metadata_collection_keys on (metadatas.address = metadata_collection_keys.metadata_address)
+            INNER JOIN auction_houses on (listings.auction_house = auction_houses.address)
+            WHERE
+                ($1 IS NULL OR metadata_collection_keys.collection_address = ANY($1))
+                AND auction_houses.treasury_mint = 'So11111111111111111111111111111111111111112'
+                AND listings.created_at >= $2
+                AND listings.created_at <= $3
```

Comment:
```suggestion
                AND listings.created_at <= $3
                AND listing.canceled_at IS NULL
                AND listing.purchase_id IS NULL
```

sorry for not getting all my CR in at once but just ran into this. Need to filter out canceled_at and purchase_id to get active listings.

Reply:
@kespinola  Correct me if I'm wrong, but if we only filter to active listings, then `COUNT( DISTINCT metadata_collection_keys.metadata_address)` will omit those NFTs which are not actively listed, which will reduce the total "supply" of NFTs in the collection, which is will result in a calculation different from what we agreed was "market cap".

I guess I can wrap in another `WITH` to get the number of NFTs in a collection without regard to listing status

Reply:
Without filtering to active listings
![image](https://user-images.githubusercontent.com/1540936/177870471-f60983ac-f425-432a-a49a-cd6d1acc1cbf.png)

versus with filtering to active listings
![image](https://user-images.githubusercontent.com/1540936/177870634-34b0d168-15b5-43a0-b636-1a408a6d5c41.png)

Reply:
First pass at this completely bogs down the query, taking ~21 seconds to complete because of computing collection sizes over the whole table.

```
WITH nft_count_table (collection, nft_count) as (
	SELECT metadata_collection_keys.collection_address AS collection, COUNT( DISTINCT metadata_collection_keys.metadata_address) as nft_count
	FROM metadata_collection_keys
	GROUP BY metadata_collection_keys.collection_address
),

market_cap_table (collection, market_cap) as (
	SELECT metadata_collection_keys.collection_address AS collection, MIN(listings.price) * MAX (nft_count_table.nft_count) AS market_cap
	FROM listings
	INNER JOIN metadatas on (listings.metadata = metadatas.address)
	INNER JOIN metadata_collection_keys on (metadatas.address = metadata_collection_keys.metadata_address)
	INNER JOIN auction_houses on (listings.auction_house = auction_houses.address)
	INNER JOIN nft_count_table on (collection = nft_count_table.collection)
	WHERE
		auction_houses.treasury_mint = 'So11111111111111111111111111111111111111112'
		AND listings.purchase_id IS NULL
		AND listings.canceled_at IS NULL
	GROUP BY metadata_collection_keys.collection_address
	ORDER BY market_cap DESC
	LIMIT 10
	OFFSET 0
) SELECT
	metadatas.name,
	metadatas.seller_fee_basis_points,
	metadatas.update_authority_address,
	metadatas.mint_address,
	metadatas.primary_sale_happened,
	metadatas.uri,
	metadatas.slot,
	metadata_jsons.description,
	metadata_jsons.image,
	metadata_jsons.category,
	metadata_jsons.model
FROM metadata_jsons, market_cap_table, metadatas
WHERE
	market_cap_table.collection = metadatas.mint_address
	AND metadatas.address = metadata_jsons.metadata_address
ORDER BY market_cap DESC
```

Reply:
With feeds there was a big performance increase by splitting the order by from the limit using another with statement. Can you test dropping the limit?

Reply:
Tried this with no benefit. Will keep poking around

Reply:
This gets it down to about 5 seconds:

```
WITH market_cap_table (collection, market_cap, nft_count, floor_price) AS (
	SELECT components.collection, components.floor_price * components.nft_count AS market_cap, components.nft_count, components.floor_price
	FROM (
		SELECT floor_prices.collection, MIN(floor_prices.floor_price) AS floor_price, COUNT( DISTINCT metadata_collection_keys.metadata_address) as nft_count
		FROM (
			SELECT metadata_collection_keys.collection_address AS collection, MIN(listings.price) AS floor_price
			FROM listings
			INNER JOIN metadatas on (listings.metadata = metadatas.address)
			INNER JOIN metadata_collection_keys on (metadatas.address = metadata_collection_keys.metadata_address)
			INNER JOIN auction_houses on (listings.auction_house = auction_houses.address)
			WHERE
				auction_houses.treasury_mint = 'So11111111111111111111111111111111111111112'
				AND listings.purchase_id IS NULL
				AND listings.canceled_at IS NULL
			GROUP BY metadata_collection_keys.collection_address
		) floor_prices
		INNER JOIN metadata_collection_keys on (metadata_collection_keys.collection_address = floor_prices.collection)
		GROUP BY floor_prices.collection
	) components
	ORDER BY market_cap DESC
	LIMIT 10
	OFFSET 0
)

SELECT
	market_cap_table.floor_price,
	market_cap_table.nft_count,
	market_cap_table.market_cap,
	metadatas.name,
	metadatas.seller_fee_basis_points,
	metadatas.update_authority_address,
	metadatas.mint_address,
	metadatas.primary_sale_happened,
	metadatas.uri,
	metadatas.slot,
	metadata_jsons.description,
	metadata_jsons.image,
	metadata_jsons.category,
	metadata_jsons.model
FROM metadatas
INNER JOIN metadata_jsons ON (metadata_jsons.metadata_address = metadatas.address)
INNER JOIN market_cap_table ON (market_cap_table.collection = metadatas.mint_address)
ORDER BY market_cap DESC
```

Reply:
with a time window this gets quicker, but still several seconds